pystapi-client/src/pystapi_client/client.py

```diff
@@ -0,0 +1,283 @@
+from typing import (
+    Any,
+    cast,
+)
+
+import httpx
+from pydantic import AnyUrl
+from stapi_pydantic import Link, Product
+from stapi_pydantic.product import ProductsCollection
+
+from pystapi_client.conformance import ConformanceClasses
+from pystapi_client.exceptions import APIError
+from pystapi_client.stapi_api_io import StapiIO
+from pystapi_client.warnings import NoConformsTo
+
+DEFAULT_LINKS = [
+    {
+        "endpoint": "/conformance",
+        "rel": "conformance",
+        "method": "GET",
+    },
+    {
+        "endpoint": "/products",
+        "rel": "products",
+        "method": "GET",
+    },
+]
+
+
+class Client:
+    """A Client for interacting with the root of a STAPI
+
+    Instances of the ``Client`` class provide a convenient way of interacting
+    with STAPI APIs that conform to the `STAPI API spec
+    <https://github.com/stapi-spec/stapi-spec>`_.
+    """
+
+    _stapi_io: StapiIO
+    _conforms_to: list[str] = []
+    _extra_fields: dict[str, Any] = {}
+
+    def __repr__(self) -> str:
+        return f"<Client {self._stapi_io.root_url}>"
+
+    @classmethod
+    def open(
+        cls,
+        url: AnyUrl,
+        headers: dict[str, str] | None = None,
+        parameters: dict[str, Any] | None = None,
+        request_modifier: Callable[[httpx.Request], httpx.Request] | None = None,
+        timeout: httpx._types.TimeoutTypes | None = None,
+    ) -> "Client":
+        """Opens a STAPI API client
+
+        Args:
+            url : The URL of a STAPI API.
+            headers : A dictionary of additional headers to use in all requests
+                made to any part of this STAPI API.
+            parameters: Optional dictionary of query string parameters to
+                include in all requests.
+            request_modifier: A callable that either modifies a `Request` instance or
+                returns a new one. This can be useful for injecting Authentication
+                headers and/or signing fully-formed requests (e.g. signing requests
+                using AWS SigV4).
+
+                The callable should expect a single argument, which will be an instance
+                of :class:`requests.Request`.
+
+                If the callable returns a `requests.Request`, that will be used.
+                Alternately, the callable may simply modify the provided request object
+                and return `None`.
+            timeout: Optional float or (float, float) tuple following the semantics
+              defined by `Requests
+              <https://requests.readthedocs.io/en/latest/api/#main-interface>`__.
+
+        Return:
+            client : A :class:`Client` instance for this STAPI API
+        """
+        client = Client()
+        client._stapi_io = StapiIO(
+            root_url=url,
+            headers=headers,
+            parameters=parameters,
+            request_modifier=request_modifier,
+            timeout=timeout,
+        )
```

Comment:
From a design perspective, are we _sure_ we want a separate IO object? Could we put the `StapiIO` methods directly on `Client`?